### PR TITLE
Clear next field when returnining list elements in queue.c

### PR DESCRIPTION
**Pull request by frozencemetery**
The ipa-otpd code occasionally removes elements from one queue,
inspects and modifies them, and then inserts them into
another (possibly identical, possibly different) queue.  When the next
pointer isn't cleared, this can result in element membership in both
queues, leading to double frees, or even self-referential elements,
causing infinite loops at traversal time.

Rather than eliminating the pattern, make it safe by clearing the next
field any time an element enters or exits a queue.

Related https://pagure.io/freeipa/issue/7262

### CODE REVIEW

**Comment by Tiboris** [comment]
Could you please add line:
`Related https://pagure.io/freeipa/issue/7262`
into commit message?

**Comment by Tiboris** [comment]
@frozencemetery nevermind I did it myself :)

**Comment by spoore1** [comment]
FYI I tested a 4.6 version of this and it seems to fix the crash.  I've run tests similar to the pagure 7262 description for many hours (12 last night, multiple tests before).  I've seen no new crashes since applying this patch.

**Comment by rcritten** [comment]
It would be really great to have a unit test to test the queueing code.

File: daemons/ipa-otpd/queue.c [hunk 1]
```diff
@@ -111,6 +111,8 @@ void otpd_queue_push(struct otpd_queue *q, struct otpd_queue_item *item)
         q->head = q->tail = item;
     else
         q->tail = q->tail->next = item;
+
+    item->next = NULL;
 }
 
 void otpd_queue_push_head(struct otpd_queue *q, struct otpd_queue_item *item)
```
Comment (flo-renaud) [inline daemons/ipa-otpd/queue.c]: In otpd_queue_push_head, we should also set item->next = NULL if the queue is currently empty.

Comment (frozencemetery) [reply daemons/ipa-otpd/queue.c]: Yes it should; good catch!  Fixed.

**Comment by frozencemetery** [comment]
Also added unit tests.

**Comment by Tiboris** [comment]
Hello @frozencemetery,
Thank you for the patch! 
Sorry for nitpicking, unfortunately line with ticket number in commit message disappeared.
Could you please add it?

**Comment by flo-renaud** [comment]
Hi @frozencemetery 
thank you for the patch and the tests, works for me. Waiting for the updated commit msg with a ref to the ticket, then I will ACK.

**Comment by frozencemetery** [comment]
Heh, sorry @Tiboris!  Fixed now.

**Comment by flo-renaud** [comment]
master:

* fe650087b0dd35b3dd23347e98647fdc526614f8 Clear next field when returnining list elements in queue.c
* ab636682c4a0274d411c702ca62a6d26ab8b7869 Add cmocka unit tests for ipa otpd queue code

**Pull request closed without merging**